nixos/modules/services/security/aesmd.nix

```diff
@@ -0,0 +1,250 @@
+{ config, pkgs, lib, ... }:
+with lib;
+let
+  cfg = config.services.aesmd;
+
+  sgx-psw = pkgs.sgx-psw.override { inherit (cfg) debug; };
+
+  configFile = with cfg.settings; pkgs.writeText "aesmd.conf" (
+    concatStringsSep "\n" (
+      optional (whitelistUrl != null) "whitelist url = ${whitelistUrl}" ++
+      optional (proxy != null) "aesm proxy = ${proxy}" ++
+      optional (proxyType != null) "proxy type = ${proxyType}" ++
+      optional (defaultQuotingType != null) "default quoting type = ${defaultQuotingType}" ++
+      # Newline at end of file
+      [ "" ]
+    )
+  );
+in
+{
+  options.services.aesmd = {
+    enable = mkEnableOption "Intel's Architectural Enclave Service Manager (AESM) for Intel SGX";
+    user = mkOption {
+      type = types.str;
+      default = "aesmd";
+      description = "Username for the AESM service.";
+    };
+    group = mkOption {
+      type = types.str;
+      default = "aesmd";
+      description = "Group for the AESM service.";
+    };
+    debug = mkOption {
+      type = types.bool;
+      default = false;
+      description = "Whether to build the PSW package in debug mode.";
+    };
+    settings = mkOption {
+      description = "AESM configuration";
+      default = { };
+      type = types.submodule {
+        options.whitelistUrl = mkOption {
+          type = with types; nullOr str;
+          default = null;
+          example = "http://whitelist.trustedservices.intel.com/SGX/LCWL/Linux/sgx_white_list_cert.bin";
+          description = "URL to retrieve authorized Intel SGX enclave signers.";
+        };
+        options.proxy = mkOption {
+          type = with types; nullOr str;
+          default = null;
+          example = "http://proxy_url:1234";
+          description = "HTTP network proxy.";
+        };
+        options.proxyType = mkOption {
+          type = with types; nullOr (enum [ "default" "direct" "manual" ]);
+          default = if (cfg.settings.proxy != null) then "manual" else null;
+          example = "default";
+          description = ''
+            Type of proxy to use. The <literal>default</literal> uses the system's default proxy.
+            If <literal>direct</literal> is given, uses no proxy.
+            A value of <literal>manual</literal> uses the proxy from
+            <option>services.aesmd.settings.proxy</option>.
+          '';
+        };
+        options.defaultQuotingType = mkOption {
+          type = with types; nullOr (enum [ "ecdsa_256" "epid_linkable" "epid_unlinkable" ]);
+          default = null;
+          example = "ecdsa_256";
+          description = "Attestation quote type.";
+        };
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    assertions = [{
+      assertion = !(config.boot.specialFileSystems."/dev".options ? "noexec");
+      message = "SGX requires exec permission for /dev";
+    }];
+
+    users.groups = {
+      "${cfg.group}" = { };
+      "sgx_prv" = { };
```

Comment:
I'm still curious about https://github.com/systemd/systemd-stable/blob/main/rules.d/50-udev-default.rules.in#L43 and if we need the "sgx" group or how "sgx_prv" is different.

Reply:
The `sgx` group is for managing access to the application enclave device, i.e., `/dev/sgx_enclave`. The `sgx_prv` group is for the provisioning device, i.e., `/dev/sgx_provision`. Using the group`sgx` for `sgx_enclave` and the group `sgx_prv` for `sgx_provision` is a convention which we should follow.

Access to the provisioning device implies access to the CPU-unique provisioning certificate key. Therefore, a user which is allowed to create application enclaves _usually_ should not be allowed to create provisioning enclaves. The AESM service is a privileged exception as it provides remote attestation.

Reply:
What is the right module in which to create the sgx group? I keep bringing this up because the existence or nonexistence of this group caused some problems in systemd startup in stage2 when there were differing bios settings.

Reply:
Before #148649, we did not create the `sgx` group for the systemd-bundled udev rule you mentioned above. This led to a warning during boot which is now resolved. We currently take care of the `sgx_prv` group in this PR only: create the group (the position you annotated) and the AESM-bundled udev rules assign permissions and create symlinks. 

As an alternative, we could take an approach similar to #148649 and create the `sgx_prv` and the respective udev rule (change group to `sgx_prv`, assign permissions `0660`) independent of having the AESM service enabled or not. I think that would make sense and would leave us with the following permissions:

```shell
# stat -c '%U %G %a %n' /dev/sgx_*
root sgx 660 /dev/sgx_enclave
root sgx_prv 660 /dev/sgx_provision
```

Anyone who should be allowed to create application enclaves joins the `sgx` group. Anyone who should be allowed to create provisioning enclaves (for example the AESM service of this PR) joins the `sgx_prv` group.

Reply:
Maybe something like https://github.com/veehaitch/nixpkgs/commit/ffed3d929c389cef33e0611a9a46fd971364f31c?